### ISSUE TITLE
Multiplayer - Users can edit at the same time

### ISSUE THREAD

**Issue by SjoenH** [equinor/MAD-VSM-WEB#92]
- [x] As a user, I want my colleagues to see the changes that I have just saved on a VSM they are also viewing. 
       https://equinor-sds-si.atlassian.net/browse/VSM-19
- [ ] #128
       https://equinor-sds-si.atlassian.net/browse/VSM-132
       https://equinor-sds-si.atlassian.net/browse/VSM-149

To-do:
- [x] #116 

# Known bugs 
## Blockers:
- [x] #123

## non-blockers
- [ ] #124
- [ ] #125
- [ ] #127
- [ ] #126

**Comment by SjoenH** [comment]
Waiting on API SignalR fixes ( @pagodo )

**Comment by SjoenH** [comment]
Skipped SignalR. Setup websocket via socket-io + nextjs api in the same docker container running alongside the webapp.
We should do some testing of multiplayer. It’s not perfect, but a big upgrade from manually refreshing the page. @ntorsen 

Code changes are in branch #106. Deployed to test now. https://web-vsm-test.radix.equinor.com

**Comment by SjoenH** [comment]
What do we do when we have two people editing one card at the same time?
Right now; It's not too bad, it does update other users sidesheets, but we get overwrite issues when someone is editing the same field.

How can we improve it?
In a perfect world, we would have proper support for editing simultaneously. Could be fun to test out. 

I'll think a bit about how we should approach this issue.

**Comment by SjoenH** [comment]
If both users have the same card open at the same time. Sync the card state asap via websocket.

**Comment by SegarJJ** [comment]
Findings from our live testing, 3 users in one VSM.

* If 2 users are editing the same description field we see that the last editing user ( user2 ) overwrites all the new text from user1.
* When two users are trying to move the same card, under certain circumstances the card can be stuck for one user.  A browser page refresh can solve the issue.
* If two user has a card open, and that same card is deleted by another user, then we could improve the message that is given to the editing user.
* Feature to show active users on a given VSM is planned but not yet implemented.

**Comment by SjoenH** [comment]
Also, deleting a vsm could trigger a message to the other users. Now it will just show error 404 that the project does not exist.

**Comment by SjoenH** [comment]
Todo:
- [ ] Add auth for socket-io endpoints

**Comment by SjoenH** [comment]
Let's move #128 out as a new feature.

**Comment by SjoenH** [comment]
Releasing what we have so far to Test and QA environments.
@equinor/mad-testers - Could you do some regression testing?

@knudsvik & @askbulle 
**Initial release of multiplayer is approaching.** I pushed what we have so far to QA environment for you to test out. 

Note: There are some known bugs:
- #124
- #125
- #127
- #126

Let me know if any of these are considered blockers for you, or if we can release what we have now and fix these later.

**Comment by askbulle** [comment]
Thorjan & I tested. Really nice feature. The bugs are should not stop us from releasing

**Comment by SjoenH** [comment]
@askbulle Noted. 
Moved the bugs into its own card https://github.com/orgs/equinor/projects/77#card-67753318  
#129 
Also, we could create an issue for the notification box if we want it to be visible when editing a card.

Doing final checks with backend @pagodo before releasing to prod. 
Radix is having some growing-pains at the moment, so release might take a little longer than usual. I'll move the card to released when finished and send you a message.

**Comment by knudsvik** [comment]
@SjoenH: @askbulle already made a separate issue for it : #130 . But we couldn't get it to show in a nice way here. (other than the number and a link)... not with a title etc.

**Comment by SjoenH** [comment]
Radix finally catched up. Multiplayer is now in Prod :muscle:
